CI: Include dSYMs in release for macOS dependencies

jcm93 opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Make sure you’ve read the contribution guidelines here: https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst -->

### Description
<!--- Describe your changes in detail. -->
<!--- If this change includes UI elements, please include screenshots. -->
Adds a wildcard to the end of the filenames to be uploaded in the final release for macOS artifacts.

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open GitHub Issue, or implements feature request -->
<!--- from the Ideas page, please link to the issue here. -->
In [releases](https://github.com/obsproject/obs-deps/releases/tag/2024-09-12), dSYMs for macOS dependencies are not being uploaded properly. Checksums are being generated, showing that the files are present. The Windows final deps upload paths include a wildcard at the end (to capture the `-PDBs` suffix), but this wildcard is not present for the macOS artifacts.

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment (hardware, OS version, etc.),-->
<!--- and the tests you ran, including how it may affect other areas of code. -->
Tested on [local fork](https://github.com/jcm93/obs-deps/releases/tag/2024-12-23).

### Types of changes
<!--- What types of changes does your PR introduce? Uncomment all that apply -->
- Bug fix (non-breaking change which fixes an issue)
<!--- - New feature (non-breaking change which adds functionality) -->
<!--- - Tweak (non-breaking change to improve existing functionality) -->
<!--- - Performance enhancement (non-breaking change which improves efficiency) -->
<!--- - Code cleanup (non-breaking change which makes code smaller or more readable) -->
<!--- - Breaking change (fix or feature that would cause existing functionality to change) -->
<!--- - Documentation (a change to documentation pages) -->

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

PatTheMav commented on .github/workflows/main.yaml:
```diff
@@ -678,6 +678,6 @@ jobs:
-            ${{ github.workspace }}/macos-*-arm64.tar.xz
-            ${{ github.workspace }}/macos-*-x86_64.tar.xz
-            ${{ github.workspace }}/macos-*-universal.tar.xz
+            ${{ github.workspace }}/macos-*-arm64*.tar.xz
+            ${{ github.workspace }}/macos-*-x86_64*.tar.xz
+            ${{ github.workspace }}/macos-*-universal*.tar.xz
```
I'd prefer a more exclusive glob expression here, e.g. `macos-*-arm64?(-dSYMs).tar.xz`.

The GitHub Actions runtime probably uses `node-glob` which explicitly supports Bash' `extglob` patterns.

PatTheMav replied:
The `softprops/action-gh-release` uses the `glob` npm package, which does.

This is not a workflow glob, but an input to a Node-based action.

jcm93 replied:
Changed; made sure the glob expression worked out-of-band and it seems to.

PatTheMav reviewed (commented):
Looks fine to me.

@jcm93 Did you run a test release on your fork to check if the glob works as intended?

jcm93 commented:
Yes; I linked it in the earlier conversation: https://github.com/jcm93/obs-deps/releases/tag/2024-12-23

PatTheMav commented:
> Yes; I linked it in the earlier conversation: https://github.com/jcm93/obs-deps/releases/tag/2024-12-23

But there's only dSYMs for Qt6 it seems?

jcm93 commented:
It looks like those are not among the downloaded artifacts for the release flow in the first place, looking at the checksums there and in the [latest official release](https://github.com/obsproject/obs-deps/releases/tag/2024-09-12). dSYMs are not present there nor are PDBs for Windows. I had actually assumed that that was intentional and there was some other reason debug symbols weren't available. If we want I can expand the scope of this to figure out why dSYMs and PDBs aren't included for the main prebuilt batch.

RytoEX commented:
> It looks like the missing dSYMs (for the generic `macos-deps-2024-12-23-arm64.tar.xz` etc.) are not among the downloaded artifacts for the release flow in the first place, looking at the checksums there and in the [latest official release](https://github.com/obsproject/obs-deps/releases/tag/2024-09-12). dSYMs are not present there nor are PDBs for the Windows counterparts (`windows-deps-2024-12-23-x64.zip` etc). I had actually assumed that that was expected and there was some other reason those debug symbols weren't available. If we want I can expand the scope of this to figure out why dSYMs and PDBs aren't included for the main prebuilt batch.

We have historically never enabled PDBs for the non-Qt Windows dependencies. That is a separate item to address (which required no longer cross-compiling many of the deps) and need not be part of this PR.

PatTheMav commented:
IMO this PR is incomplete if only the Qt6 dSYMs are made available in a release (as all the other dSYMs are generated as artefacts). Because FFmpeg and main dependencies are smushed together into a new archive, you'd probably have to extend the workflow step to do the same for the dSYM archives as well.

jcm93 commented:
I'll take a look at getting dSYMs for everything else.

RytoEX commented:
> IMO this PR is incomplete if only the Qt6 dSYMs are made available in a release (as all the other dSYMs are generated as artefacts). Because FFmpeg and main dependencies are smushed together into a new archive, you'd probably have to extend the workflow step to do the same for the dSYM archives as well.

My main point about the Windows devs is that we do not even _generate_ PDBs for the FFmpeg dependencies on Windows yet. Addressing that here would be out of scope. If PDBs for everything else are already generated, but not included, then yes, it is fine to address that here.

PatTheMav commented:
> > IMO this PR is incomplete if only the Qt6 dSYMs are made available in a release (as all the other dSYMs are generated as artefacts). Because FFmpeg and main dependencies are smushed together into a new archive, you'd probably have to extend the workflow step to do the same for the dSYM archives as well.
> 
> My main point about the Windows deps is that we do not even _generate_ PDBs for the FFmpeg dependencies on Windows yet. Addressing that here would be out of scope. If PDBs for everything else are already generated, but not included, then yes, it is fine to address that here.

Yep the PDBs are a different matter entirely. For macOS all dependencies flavours generate dSYMs already though.

jcm93 commented:
Updated; added a new step that packages the debug symbols, since it uses a separate glob expression. Test release is [here](https://github.com/jcm93/obs-deps/releases/tag/2025-01-20).

PatTheMav reviewed (commented):
There is no need for an additional step, this should do just fine:

```Bash
for arch in arm64 x86_64 universal; do
  _temp=$(mktemp -d)
  pushd "${_temp}" > /dev/null
  for artifact in ${GITHUB_WORKSPACE}/**/macos-@(deps|ffmpeg)-!(qt6*)-${arch}?(-dSYMs).*; do
    case ${artifact} in
      *.zip) unzip -o ${artifact} > /dev/null ;;
      *.tar.xz) XZ_OPT=-T0 tar -xvJf ${artifact} ;;
      *.tar.gz) tar -xvzf ${artifact} ;;
    esac
  done
  XZ_OPT=-T0 tar -cvJf macos-deps-${{ steps.metadata.outputs.version }}-${arch}.tar.xz -- !(*.dSYM)
  XZ_OPT=-T0 tar -cvJf macos-deps-${{ steps.metadata.outputs.version }}-${arch}-dSYMs.tar.xz -- *.dSYM
  mv macos-deps-${{ steps.metadata.outputs.version }}-${arch}?(-dSYMs).tar.xz ${GITHUB_WORKSPACE}
  popd > /dev/null
done
```

jcm93 commented:
Alright yeah that seems to have worked. Pushed; test release [here](https://github.com/jcm93/obs-deps/releases/tag/2025-01-20).

RytoEX commented on .github/workflows/main.yaml:
```diff
@@ -640,18 +640,16 @@ jobs:
           for arch in arm64 x86_64 universal; do
             _temp=$(mktemp -d)
             pushd "${_temp}" > /dev/null
-
```
Can probably leave this blank line intact.

jcm93 replied:
Addressed.

RytoEX commented on .github/workflows/main.yaml:
```diff
@@ -640,18 +640,16 @@ jobs:
                 *.tar.gz) tar -xvzf ${artifact} ;;
               esac
             done
-
-            XZ_OPT=-T0 tar -cvJf macos-deps-${{ steps.metadata.outputs.version }}-${arch}.tar.xz -- *
-            mv macos-deps-${{ steps.metadata.outputs.version }}-${arch}.tar.xz ${GITHUB_WORKSPACE}
-
```
Same for these blank lines.

jcm93 replied:
Addressed.

PatTheMav reviewed (approved):
Looks good to me - obviously depends on the artefact contents staying as-is, but the entire CI infrastructure depends on every file fulfilling some expectation like that. 😅
